storage: retry AdminSplit on seeing roachpb.ConflictUpdatingRangeDesc

vivekmenezes opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10728)
<!-- Reviewable:end -->

bdarnell commented:
Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/kv/dist_sender.go, line 936 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWlAHo9S6ihoEr69wiD:-KWlAHo9S6ihoEr69wiE:b-q83hxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/49acfd29bc1bced26e12168941ca07029068c04b/pkg/kv/dist_sender.go#L936)):*
> ```Go
> 			reply, pErr = ds.divideAndSendBatchToRanges(ctx, ba, intersected, isFirst)
> 			return response{reply: reply, pErr: pErr}
> 		case *roachpb.ConflictUpdatingRangeDesc:
> ```

This is not the right place to retry this error. The whole reason we pass the descriptor into AdminSplit is because if the descriptor changes out from under us, we may no longer want to do the split (someone else may have already split the range down to size). 

Move the retry logic up to the higher level: in splitQueue, if there is a conflict, we should MaybeAdd(repl) to go back to the beginning and decide whether or not we want to retry, and if so where the new split point should be. In user-directed splits like the SQL `SPLIT AT` command, we probably do want to retry on conflicting updates until the split succeeds. 

---

*[pkg/roachpb/errors.proto, line 215 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWlA46GZTKhs8l-SyH8:-KWlA46GZTKhs8l-SyH9:bxzh397) ([raw file](https://github.com/cockroachdb/cockroach/blob/49acfd29bc1bced26e12168941ca07029068c04b/pkg/roachpb/errors.proto#L215)):*
> ```protobuf
> // when it conflicts with some other process that updates the
> // range descriptors.
> message ConflictUpdatingRangeDesc {
> ```

Our other error types all end in `Error`. 

---

*[pkg/storage/replica_command.go, line 2447 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWl9wYElMUJUm7vTvsr:-KWl9wYFtsrwumOpwCYS:b-ymmjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/49acfd29bc1bced26e12168941ca07029068c04b/pkg/storage/replica_command.go#L2447)):*
> ```Go
> 	}); err != nil {
> 		if err == errMsgConflictUpdatingRangeDesc {
> 			return reply, roachpb.NewError(&roachpb.ConflictUpdatingRangeDesc{})
> ```

roachpb.ConflictUpdatingRangeDesc implements Error; why can't we just use it directly instead of `errMsgConflictUpdatingRangeDesc`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 4 of 11 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/kv/dist_sender.go, line 936 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWlAHo9S6ihoEr69wiD:-KWn3gEhicPV9CCgEQbD:b5ykc72) ([raw file](https://github.com/cockroachdb/cockroach/blob/49acfd29bc1bced26e12168941ca07029068c04b/pkg/kv/dist_sender.go#L936)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This is not the right place to retry this error. The whole reason we pass the descriptor into AdminSplit is because if the descriptor changes out from under us, we may no longer want to do the split (someone else may have already split the range down to size). 
> 
> Move the retry logic up to the higher level: in splitQueue, if there is a conflict, we should MaybeAdd(repl) to go back to the beginning and decide whether or not we want to retry, and if so where the new split point should be. In user-directed splits like the SQL `SPLIT AT` command, we probably do want to retry on conflicting updates until the split succeeds. 
</details>
I've updated the comments in SplitQueue explaining the rationale for using AdminSplit() and an AdminSplitRequest{} for the two cases where a split is requested. I believe the AdminSplit() use case in SplitQueue can benefit from retrying of the split on seeing a conflict. If it does see a conflict with an AdminSplit on the same key, it will see a different error `range is already split at key` which gets returned.

---

*[pkg/roachpb/errors.proto, line 215 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWlA46GZTKhs8l-SyH8:-KWn4SEYH2yM2lmur1Lc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/49acfd29bc1bced26e12168941ca07029068c04b/pkg/roachpb/errors.proto#L215)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Our other error types all end in `Error`. 
</details>
Done.

---

*[pkg/storage/replica_command.go, line 2447 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWl9wYElMUJUm7vTvsr:-KWn4U0vNp_rizswXVQF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/49acfd29bc1bced26e12168941ca07029068c04b/pkg/storage/replica_command.go#L2447)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> roachpb.ConflictUpdatingRangeDesc implements Error; why can't we just use it directly instead of `errMsgConflictUpdatingRangeDesc`?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 10 files at r1, 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/kv/dist_sender.go, line 936 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWlAHo9S6ihoEr69wiD:-KWnBfASOl-VNcuBf1-y:b-950jqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/49acfd29bc1bced26e12168941ca07029068c04b/pkg/kv/dist_sender.go#L936)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I've updated the comments in SplitQueue explaining the rationale for using AdminSplit() and an AdminSplitRequest{} for the two cases where a split is requested. I believe the AdminSplit() use case in SplitQueue can benefit from retrying of the split on seeing a conflict. If it does see a conflict with an AdminSplit on the same key, it will see a different error `range is already split at key` which gets returned.
</details>
+1 to @bdarnell's comment that retrying at the dist sender is the wrong level. I didn't quite understand your reply to that.

Retrying at the client will also allow you to avoid the extra structured error you added here.

---

*[pkg/storage/client_split_test.go, line 304 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWn79qMUarXOhZ2c8oS:-KWn79qMUarXOhZ2c8oT:b-dj44qc) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd810d8d72e3c57d86c6a21e968e988e09deae90/pkg/storage/client_split_test.go#L304)):*
> ```Go
> 			// split key is outside of the bounds for the range.
> 			expected := strings.Join([]string{
> 				"conflict updating range descriptor",
> ```

why a literal string? you should probably match the structured error?

---

*[pkg/storage/replica_command.go, line 2389 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWnCP_AipOwst7vnEru:-KWnCP_AipOwst7vnErv:b-a1e36s) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd810d8d72e3c57d86c6a21e968e988e09deae90/pkg/storage/replica_command.go#L2389)):*
> ```Go
> 				// left range descriptor is read outside of the transaction.
> 				if _, ok := err.(*roachpb.ConditionFailedError); ok {
> 					return &roachpb.ConflictUpdatingRangeDescError{}
> ```

@bdarnell suggested trying above this level, but I think it might be OK to do it here, too.

---

*[pkg/storage/replica_command.go, line 2442 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWn7GKYQcpTi8WgJ0Na:-KWn7GKYQcpTi8WgJ0Nb:b-fzfv8u) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd810d8d72e3c57d86c6a21e968e988e09deae90/pkg/storage/replica_command.go#L2442)):*
> ```Go
> 		return nil
> 	}); err != nil {
> 		if pErr, ok := err.(*roachpb.ConflictUpdatingRangeDescError); ok {
> ```

`pErr` is what we name things of type `roachpb.Error`, which this is not.

---

*[pkg/storage/split_queue.go, line 106 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWnCnwUzeezH8zw7h4F:-KWnCnwUzeezH8zw7h4G:b9q83ug) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd810d8d72e3c57d86c6a21e968e988e09deae90/pkg/storage/split_queue.go#L106)):*
> ```Go
> 		for _, splitKey := range splitKeys {
> 			// Use AdminSplit() here because the zone config stipulates
> 			// splitting at a specific key independent of which replica it
> ```

s/replica/range/ unless I'm misunderstanding.

---

*[pkg/storage/split_queue.go, line 108 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWnCvNop7nWlwotgJ8s:-KWnCvNop7nWlwotgJ8t:b-wf535a) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd810d8d72e3c57d86c6a21e968e988e09deae90/pkg/storage/split_queue.go#L108)):*
> ```Go
> 			// splitting at a specific key independent of which replica it
> 			// lies on (the specific replica desc passed in can change before
> 			// the command is applied). If a range was already split at this
> ```

s/a/any/

---

*[pkg/storage/split_queue.go, line 126 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWnCzefXSBCIF_uTSM1:-KWnCzefXSBCIF_uTSM2:b-gkvzze) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd810d8d72e3c57d86c6a21e968e988e09deae90/pkg/storage/split_queue.go#L126)):*
> ```Go
> 		log.Infof(ctx, "splitting size=%d max=%d", size, zone.RangeMaxBytes)
> 		// Do not use the AdminSplit() command because we are attempting to
> 		// split a specific replica at a key. If this replica has already been
> ```

i don't understand how these comments address the FIXME that previously occupied this space. ISTM that they do the exact same thing (since the replica is the lease holer, the AdminSplit above, via dist sender + local call optimization will result in exactly the same call to the replica).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 9 of 11 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/storage/split_queue.go, line 126 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWnCzefXSBCIF_uTSM1:-KWqb88ONJG1s9nujP1U:bfg0epm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd810d8d72e3c57d86c6a21e968e988e09deae90/pkg/storage/split_queue.go#L126)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i don't understand how these comments address the FIXME that previously occupied this space. ISTM that they do the exact same thing (since the replica is the lease holer, the AdminSplit above, via dist sender + local call optimization will result in exactly the same call to the replica).
</details>
But this calls directly to the replica while the preceding one goes through the DistSender. This gives them different semantics WRT retries on conflict, although that is subtle and indirect and doesn't make me feel any better about doing that retry in DistSender instead of either at the various callers of AdminSplit or internal to AdminSplit itself. 

This should really be calling `Replica.AdminSplit()` directly (instead of going through `Replica.Send`. Or we should add a RangeDescriptor to AdminSplitRequest to unify the two interfaces) so it can pass in the descriptor we had when we made the decision to split. When there is a conflict, control must be returned here so we can see if the split is still necessary. 

---

*[pkg/storage/split_queue.go, line 125 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWqaufYHtwgat-3p_hn:-KWqaufYHtwgat-3p_ho:ba6xsim) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c171724c138b8c197324d9e3a35078bdcc3a8da/pkg/storage/split_queue.go#L125)):*
> ```Go
> 	if float64(size)/float64(zone.RangeMaxBytes) > 1 {
> 		log.Infof(ctx, "splitting size=%d max=%d", size, zone.RangeMaxBytes)
> 		// We cannot use the AdminSplit() command here because it can only be
> ```

What distinction are you trying to make here? You're sending an AdminSplitRequest which will be processed by `Replica.AdminSplit()`. What is "the AdminSplit() command"? (we normally use "command" to refer to raft commands, which AdminSplit is not)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 8 of 11 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/kv/dist_sender.go, line 936 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWlAHo9S6ihoEr69wiD:-KWnInOuTEr3u4QRNAce:b-b6jn6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/49acfd29bc1bced26e12168941ca07029068c04b/pkg/kv/dist_sender.go#L936)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> +1 to @bdarnell's comment that retrying at the dist sender is the wrong level. I didn't quite understand your reply to that.
> 
> Retrying at the client will also allow you to avoid the extra structured error you added here.
</details>
I believe there is a concern that under certain circumstances a retry might be unwarranted, in particular in SplitQueue, but there is a reason why AdminSplit() is used for splits due to zone config, and AdminSplitRequest is used for splits due to range size. 

---

*[pkg/storage/replica_command.go, line 2442 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWn7GKYQcpTi8WgJ0Na:-KWnK9kgmm55uDeARuFl:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd810d8d72e3c57d86c6a21e968e988e09deae90/pkg/storage/replica_command.go#L2442)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `pErr` is what we name things of type `roachpb.Error`, which this is not.
</details>
fixed

---

*[pkg/storage/split_queue.go, line 106 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWnCnwUzeezH8zw7h4F:-KWnKOosz96cl6CVgxz5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd810d8d72e3c57d86c6a21e968e988e09deae90/pkg/storage/split_queue.go#L106)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/replica/range/ unless I'm misunderstanding.
</details>
Done.

---

*[pkg/storage/split_queue.go, line 108 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWnCvNop7nWlwotgJ8s:-KWnKXt9Xq5N2V7QybCs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd810d8d72e3c57d86c6a21e968e988e09deae90/pkg/storage/split_queue.go#L108)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/a/any/
</details>
Done.

---

*[pkg/storage/split_queue.go, line 126 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWnCzefXSBCIF_uTSM1:-KWnRZH2e58tqtbxJAFo:b-ky73q9) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd810d8d72e3c57d86c6a21e968e988e09deae90/pkg/storage/split_queue.go#L126)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> But this calls directly to the replica while the preceding one goes through the DistSender. This gives them different semantics WRT retries on conflict, although that is subtle and indirect and doesn't make me feel any better about doing that retry in DistSender instead of either at the various callers of AdminSplit or internal to AdminSplit itself. 
> 
> This should really be calling `Replica.AdminSplit()` directly (instead of going through `Replica.Send`. Or we should add a RangeDescriptor to AdminSplitRequest to unify the two interfaces) so it can pass in the descriptor we had when we made the decision to split. When there is a conflict, control must be returned here so we can see if the split is still necessary. 
</details>
I've reworded the comments and also call Replica.AdminSplit() directly here. I do not catch the `roachpb.ConflictUpdatingRangeDescError` here because it's not really worth retrying that here. If a conflict really occurred we can either assume that the right thing happened through another split, or we'll retry again at another time when the replica gets queued again.

---

*[pkg/storage/split_queue.go, line 125 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWqaufYHtwgat-3p_hn:-KWtP-s3zT_hzJJ4ILSi:b-cz0p3w) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c171724c138b8c197324d9e3a35078bdcc3a8da/pkg/storage/split_queue.go#L125)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> What distinction are you trying to make here? You're sending an AdminSplitRequest which will be processed by `Replica.AdminSplit()`. What is "the AdminSplit() command"? (we normally use "command" to refer to raft commands, which AdminSplit is not)
</details>
I've replaced AdminSplit() with db.AdminSplit(), saying that we can't use db.AdminSplit() here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 10 files at r1, 1 of 2 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/kv/dist_sender.go, line 936 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWlAHo9S6ihoEr69wiD:-KWtW1S1VrH5BwhQdvRE:b-j9dkes) ([raw file](https://github.com/cockroachdb/cockroach/blob/49acfd29bc1bced26e12168941ca07029068c04b/pkg/kv/dist_sender.go#L936)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I believe there is a concern that under certain circumstances a retry might be unwarranted, in particular in SplitQueue, but there is a reason why AdminSplit() is used for splits due to zone config, and AdminSplitRequest is used for splits due to range size. 
</details>
What reason is that? I've opened a separate PR to disentangle that discussion from this PR (which is about retrying the split): https://github.com/cockroachdb/cockroach/pull/10787.

---

*[pkg/storage/split_queue.go, line 125 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWqaufYHtwgat-3p_hn:-KWtU5C5jaGhGEmjXhCw:bhe330e) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c171724c138b8c197324d9e3a35078bdcc3a8da/pkg/storage/split_queue.go#L125)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I've replaced AdminSplit() with db.AdminSplit(), saying that we can't use db.AdminSplit() here.
</details>
I don't see any reason this can't call `r.AdminSplit()` in both cases. Since this is not really what this PR is about, I've extracted this change into a separate PR: https://github.com/cockroachdb/cockroach/pull/10787.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/storage/split_queue.go, line 105 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWtaU2GiAMKVMrb5H6R:-KWtaU2GiAMKVMrb5H6S:bmofafr) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e70a432fd7e8c8293181021f5bca8e2c1542185/pkg/storage/split_queue.go#L105)):*
> ```Go
> 		log.Infof(ctx, "splitting at keys %v", splitKeys)
> 		for _, splitKey := range splitKeys {
> 			// Use db.AdminSplit() here because the zone config stipulates
> ```

sorry, one more cook... I think this should also call `r.AdminSplit()`, like the case below. I don't think the theoretical extra chances of success justify this confusing code. The queues are async processors after all, so as long as the right thing eventually happens, that's good enough.
The comment also doesn't help, because it doesn't contrast and explain the difference between `db.AdminSplit` and `r.AdminSplit` (what Ben explained below - the role of the DistSender). In fact, for me, the comment as it is worded now suggests that `db.AdminSplit` is *more likely* to return an error.

---

*[pkg/storage/split_queue.go, line 132 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWtc17ivlzgTphMaAvR:-KWtc17ivlzgTphMaAvS:b43vusy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e70a432fd7e8c8293181021f5bca8e2c1542185/pkg/storage/split_queue.go#L132)):*
> ```Go
> 		// pass it into AdminSplit().
> 		//
> 		// Note: another split can occur between the time the range size is
> ```

explain what happens then in this comment

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Ready for another review.

---

Review status: 9 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/kv/dist_sender.go, line 936 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWlAHo9S6ihoEr69wiD:-KWuKbJbkweFgwXSBzEe:b-k4a8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/49acfd29bc1bced26e12168941ca07029068c04b/pkg/kv/dist_sender.go#L936)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> What reason is that? I've opened a separate PR to disentangle that discussion from this PR (which is about retrying the split): https://github.com/cockroachdb/cockroach/pull/10787.
</details>
I've rebased on top of that PR

---

*[pkg/storage/client_split_test.go, line 304 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWn79qMUarXOhZ2c8oS:-KWuKeZFmxR-dO-P330O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd810d8d72e3c57d86c6a21e968e988e09deae90/pkg/storage/client_split_test.go#L304)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why a literal string? you should probably match the structured error?
</details>
Done.

---

*[pkg/storage/split_queue.go, line 125 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWqaufYHtwgat-3p_hn:-KWuKzc4vBO-s0jkMzWS:b-k73z76) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c171724c138b8c197324d9e3a35078bdcc3a8da/pkg/storage/split_queue.go#L125)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I don't see any reason this can't call `r.AdminSplit()` in both cases. Since this is not really what this PR is about, I've extracted this change into a separate PR: https://github.com/cockroachdb/cockroach/pull/10787.
</details>
thanks

---

*[pkg/storage/split_queue.go, line 105 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWtaU2GiAMKVMrb5H6R:-KWuKtrs7IdW3JNnFhyF:b5gynu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e70a432fd7e8c8293181021f5bca8e2c1542185/pkg/storage/split_queue.go#L105)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> sorry, one more cook... I think this should also call `r.AdminSplit()`, like the case below. I don't think the theoretical extra chances of success justify this confusing code. The queues are async processors after all, so as long as the right thing eventually happens, that's good enough.
> The comment also doesn't help, because it doesn't contrast and explain the difference between `db.AdminSplit` and `r.AdminSplit` (what Ben explained below - the role of the DistSender). In fact, for me, the comment as it is worded now suggests that `db.AdminSplit` is *more likely* to return an error.
</details>

that was just merged in : #10787

---

*[pkg/storage/split_queue.go, line 132 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KWtc17ivlzgTphMaAvR:-KWuL2QAf8-cr-GJ3hGN:b32vzs) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e70a432fd7e8c8293181021f5bca8e2c1542185/pkg/storage/split_queue.go#L132)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> explain what happens then in this comment
</details>
NA

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-:-KWuLxPA2-ULudFm5sjN:b-qhjwo0)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
I still object to doing this at the DistSender level. How about retrying in `Replica.AdminSplit` if a key was specified in the request? (if no key was specified, we can return the error to the caller). In this case we can even get rid of the new proto error since the retry can be handled locally. 

---

Review status: 9 of 10 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-:-KWvf6Fzp5WpjxS3C-4m:blzptzk)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Ah ha it occurred to me that when we retry the request at the replica, if
the replica no longer holds the key, it will get retried at the dist sender
level. So yes doing this at the replica level will work.

On Sat, Nov 19, 2016, 3:23 AM Ben Darnell <notifications@github.com> wrote:

> I still object to doing this at the DistSender level. How about retrying
> in Replica.AdminSplit if a key was specified in the request? (if no key
> was specified, we can return the error to the caller). In this case we can
> even get rid of the new proto error since the retry can be handled locally.
> ------------------------------
>
> Review status: 9 of 10 files reviewed at latest revision, 6 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-:-KWvf6Fzp5WpjxS3C-4m:blzptzk>*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/10728#issuecomment-261700963>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBPwpkRj3FWtjet0Ys6ogQ8xex9l7ks5q_rHogaJpZM4Kz9JT>
> .
>

vivekmenezes committed:
storage: retry AdminSplit on roachpb.ConditionFailedError

This allows an internal error to be retried instead of it needing to
be retried on the outside.

vivekmenezes commented:
I've implemented the retry at the AdminSplit command level. Ready for another review. Thanks

---

Review status: 3 of 5 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-:-KX3Xouk1yy3cKzMa7pV:beqjw55)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r5, 7 of 7 files at r6.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_split_test.go, line 304 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KX3z0aw85VICkWB-KL3:-KX3z0aw85VICkWB-KL4:b-f3m729) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ddb9518a3c6d25a0a9488357ea9b6cb84d680f1/pkg/storage/client_split_test.go#L304)):*
> ```Go
> 			// split key is outside of the bounds for the range.
> 			expected := strings.Join([]string{
> 				(&roachpb.ConditionFailedError{}).String(),
> ```

is it actually possible for this error to leak out of the store? if so, why did the error handling in sql/split.go and storage.replica_queue_test.go go away?

---

*[pkg/storage/replica_command.go, line 2437 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KX3zODDuE_IkGYOpvk_:-KX3zODDuE_IkGYOpvka:b2t5eq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ddb9518a3c6d25a0a9488357ea9b6cb84d680f1/pkg/storage/replica_command.go#L2437)):*
> ```Go
> 			// key was provided to the AdminSplit command.
> 			if _, ok := err.(*roachpb.ConditionFailedError); ok && keyProvided {
> 				desc = r.Desc()
> ```

this seems pretty sketchy - why even take a descriptor argument if we're going to just grab it off the replica?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r4, 7 of 7 files at r6.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_split_test.go, line 304 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KX3z0aw85VICkWB-KL3:-KX4w0N3WCgT8vNRk0JA:b-dp09tg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ddb9518a3c6d25a0a9488357ea9b6cb84d680f1/pkg/storage/client_split_test.go#L304)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is it actually possible for this error to leak out of the store? if so, why did the error handling in sql/split.go and storage.replica_queue_test.go go away?
</details>
It should only be able to leak out if no split key is given, which is not the case in this test.

---

*[pkg/storage/replica_command.go, line 2302 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KX4wx2Fa1PUXlzQCzq1:-KX4wx2Fa1PUXlzQCzq2:bke8770) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ddb9518a3c6d25a0a9488357ea9b6cb84d680f1/pkg/storage/replica_command.go#L2302)):*
> ```Go
> 	keyProvided := len(args.SplitKey) != 0
> 
> 	for retryable := retry.StartWithCtx(ctx, retry.Options{}); retryable.Next(); {
> ```

Use `base.DefaultRetryOptions()`, not `retry.Options{}`.

---

*[pkg/storage/replica_command.go, line 2437 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KX3zODDuE_IkGYOpvk_:-KX50kJ1kCCTdZkye2Fk:b-jpz8qi) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ddb9518a3c6d25a0a9488357ea9b6cb84d680f1/pkg/storage/replica_command.go#L2437)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this seems pretty sketchy - why even take a descriptor argument if we're going to just grab it off the replica?
</details>
The two use cases for this method are diverging: There's the split-by-size case, which passes a descriptor and wants a single attempt with that value as the CPut expected value, and the split-at-key case which passes a key and wants multiple attempts (each using the current r.Desc() as its expected value).

It might clean things up to leave AdminSplit as it was and introduce a new AdminSplitAtKey method, which would not take a descriptor but would contain the retry loop and calls to `r.Desc()` on each iteration. Or it might not. This is fine with me too but probably needs a comment about why we're resetting the descriptor. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-:-KX52a5YMOImGVCkgQUy:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_split_test.go, line 304 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KX3z0aw85VICkWB-KL3:-KX5VlbqrqGaSos4tbqw:b-54rgm1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ddb9518a3c6d25a0a9488357ea9b6cb84d680f1/pkg/storage/client_split_test.go#L304)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> It should only be able to leak out if no split key is given, which is not the case in this test.
</details>
Right, so this should be removed.

---

*[pkg/storage/replica_command.go, line 2437 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KX3zODDuE_IkGYOpvk_:-KX5VHyhWWyqLRPSAj_B:bas15yr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ddb9518a3c6d25a0a9488357ea9b6cb84d680f1/pkg/storage/replica_command.go#L2437)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The two use cases for this method are diverging: There's the split-by-size case, which passes a descriptor and wants a single attempt with that value as the CPut expected value, and the split-at-key case which passes a key and wants multiple attempts (each using the current r.Desc() as its expected value).
> 
> It might clean things up to leave AdminSplit as it was and introduce a new AdminSplitAtKey method, which would not take a descriptor but would contain the retry loop and calls to `r.Desc()` on each iteration. Or it might not. This is fine with me too but probably needs a comment about why we're resetting the descriptor. 
</details>
Splitting the two methods SGTM.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 3 of 7 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/storage/client_split_test.go, line 304 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KX3z0aw85VICkWB-KL3:-KX6UQY_Fs4GBH0nKlvn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ddb9518a3c6d25a0a9488357ea9b6cb84d680f1/pkg/storage/client_split_test.go#L304)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Right, so this should be removed.
</details>
Done.

---

*[pkg/storage/replica_command.go, line 2302 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KX4wx2Fa1PUXlzQCzq1:-KX6UJwdwARjXD_tG-pZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ddb9518a3c6d25a0a9488357ea9b6cb84d680f1/pkg/storage/replica_command.go#L2302)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Use `base.DefaultRetryOptions()`, not `retry.Options{}`.
</details>
Done.

---

*[pkg/storage/replica_command.go, line 2437 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KX3zODDuE_IkGYOpvk_:-KX6UOitY11n8aIySLYo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ddb9518a3c6d25a0a9488357ea9b6cb84d680f1/pkg/storage/replica_command.go#L2437)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Splitting the two methods SGTM.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 9 of 9 files at r7.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica_command.go, line 2281 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KX6XeYboKJgLFdfSTpt:-KX6XX12BtydyNoZBdUd:b-qcaai7) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdea31d65dfeefb8586d951f64962b117ade2edb/pkg/storage/replica_command.go#L2281)):*
> ```Go
> 	ctx context.Context, args roachpb.AdminSplitRequest,
> ) (roachpb.AdminSplitResponse, *roachpb.Error) {
> 	var reply roachpb.AdminSplitResponse
> ```

nit: remove this and use literal `roachpb.AdminSplitResponse{}` at the error returns.

---

*[pkg/storage/replica_command.go, line 2294 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KX6XZzhXT-t2Wl9uXV-:-KX6XZzhXT-t2Wl9uXV0:bo7s1ds) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdea31d65dfeefb8586d951f64962b117ade2edb/pkg/storage/replica_command.go#L2294)):*
> ```Go
> 		}
> 	}
> 	return reply, nil
> ```

Uh, this doesn't seem right. You probably want to return `reply, roachpb.NewError(ctx.Err())` here.

---

*[pkg/storage/replica_command.go, line 2453 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KX6Y-xxEQgldl_o7vtX:-KX6Y-xxEQgldl_o7vtY:b-ujnbp0) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdea31d65dfeefb8586d951f64962b117ade2edb/pkg/storage/replica_command.go#L2453)):*
> ```Go
> 		// ConditionFailedError in the error detail so that the command can be
> 		// retried.
> 		if failure, ok := err.(*roachpb.ConditionFailedError); ok {
> ```

nit: there is no need for the `failure` binding at all:
```
if _, ok := err.(*roachpb.ConditionFailedError); ok {
  return reply, roachpb.NewError(err)
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-:-KX6YYozdtFxOp0FvNbA:bnfp4nl)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 7 of 8 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/storage/replica_command.go, line 2281 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KX6XeYboKJgLFdfSTpt:-KX6_N3R5fuD6IlvlLWU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdea31d65dfeefb8586d951f64962b117ade2edb/pkg/storage/replica_command.go#L2281)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: remove this and use literal `roachpb.AdminSplitResponse{}` at the error returns.
</details>
Done.

---

*[pkg/storage/replica_command.go, line 2294 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KX6XZzhXT-t2Wl9uXV-:-KX6Zl1TfLzUCO1ypUtT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdea31d65dfeefb8586d951f64962b117ade2edb/pkg/storage/replica_command.go#L2294)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Uh, this doesn't seem right. You probably want to return `reply, roachpb.NewError(ctx.Err())` here.
</details>
Done.

---

*[pkg/storage/replica_command.go, line 2453 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728#-KX6Y-xxEQgldl_o7vtX:-KX6ZsUjC9ebJcODK11T:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdea31d65dfeefb8586d951f64962b117ade2edb/pkg/storage/replica_command.go#L2453)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: there is no need for the `failure` binding at all:
> ```
> if _, ok := err.(*roachpb.ConditionFailedError); ok {
>   return reply, roachpb.NewError(err)
> }
> ```
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10728)*
<!-- Sent from Reviewable.io -->
